src/build_vehicle_gui.cpp

```diff
@@ -1232,7 +1609,11 @@ struct BuildVehicleWindow : Window {
 	bool FilterSingleEngine(EngineID eid)
 	{
 		CargoID filter_type = this->cargo_filter[this->cargo_filter_criteria];
-		return (filter_type == CF_ANY || CargoFilter(&eid, filter_type));
+		return (filter_type == CF_ANY || CargoFilterFunc(&eid, filter_type));
+	}
+
+	bool IsSmartMenu() {
+		return (window_desc->nwid_length == sizeof _nested_build_vehicle_smart_widgets / sizeof _nested_build_vehicle_smart_widgets[0]);
```

Comment:
I'm not sure I like this way of testing, but anyway you can use the `lengthof()` macro that does exactly this, gets the static length of an array.

Reply:
I completely agree

Reply:
Actually, I initially thought of making a separate menu in a separate file. But much of it would be a copy. And to support both options would be terrible.